File: mapf_ros2_ws/src/mapf_ros/mapf_ros/core/agent.py
"""
agent.py — Agent Model (Python / ROS2-ready)
"""

from __future__ import annotations
from enum import Enum
from typing import List, Optional


class AgentState(str, Enum):
    IDLE     = "IDLE"
    PLANNING = "PLANNING"
    MOVING   = "MOVING"
    WAITING  = "WAITING"
    ARRIVED  = "ARRIVED"
    STUCK    = "STUCK"


class Agent:
    """Represents one autonomous agent in the MAPF system.

    Path is a list of {"x", "y", "t"} dicts produced by Space-Time A*.
    The agent tracks its logical (grid) position and computes the next
    waypoint for the physical controller.
    """

    def __init__(self, agent_id: int, start: dict, goal: dict, priority: int) -> None:
        self.id       = agent_id
        self.start    = dict(start)
        self.goal     = dict(goal)
        self.priority = priority

        # Current logical grid position
        self.grid_x: int = start["x"]
        self.grid_y: int = start["y"]

        # Planned space-time path: [{"x","y","t"}, ...]
        self.path: List[dict] = []
        self.path_index: int  = 0

        self.state: AgentState = AgentState.IDLE
        self.start_time: int   = 0

        # Stats
        self.steps_taken: int      = 0
        self.wait_steps:  int      = 0
        self.replan_count: int     = 0
        self.conflicts_involved: int = 0

        # For the physical controller: target waypoint in world coords
        # Updated by the controller when waypoint is reached
        self._current_waypoint_idx: int = 0

    # ── Path assignment ────────────────────────────────────────────────────

    def assign_path(self, path: Optional[List[dict]]) -> None:
        self.path       = path or []
        self.path_index = 0
        if not self.path:
            self.state = AgentState.STUCK
        else:
            self.state = AgentState.MOVING

    def has_remaining_path(self) -> bool:
        return self.path_index < len(self.path) - 1

    # ── Logical step (advances grid position per timestep) ─────────────────

    def step(self, current_t: int) -> None:
        if self.state in (AgentState.ARRIVED, AgentState.STUCK):
            return
        if not self.path:
            self.state = AgentState.STUCK
            return

        idx = self.path_index
        if idx >= len(self.path):
            self.state = AgentState.ARRIVED
            return

        entry = self.path[idx]
        if entry["t"] > current_t:
            self.state = AgentState.WAITING
            return

        moved = (entry["x"] != self.grid_x or entry["y"] != self.grid_y)
        self.grid_x = entry["x"]
        self.grid_y = entry["y"]

        if moved:
            self.steps_taken += 1
        else:
            self.wait_steps += 1

        if self.path_index + 1 < len(self.path):
            self.path_index += 1

        if self.grid_x == self.goal["x"] and self.grid_y == self.goal["y"]:
            self.state = AgentState.ARRIVED
        else:
            self.state = AgentState.MOVING if moved else AgentState.WAITING

    # ── Physical controller waypoint helpers ───────────────────────────────

    def next_world_waypoint(self) -> Optional[tuple]:
        """Return (world_x, world_y) of the next waypoint in the plan.

        Returns None if the agent has completed its path.
        """
        if not self.path:
            return None
        idx = min(self._current_waypoint_idx, len(self.path) - 1)
        p   = self.path[idx]
        return float(p["x"]) + 0.5, float(p["y"]) + 0.5

    def advance_waypoint(self) -> None:
        """Called by controller when current waypoint is reached."""
        if self._current_waypoint_idx < len(self.path) - 1:
            self._current_waypoint_idx += 1

    @property
    def progress(self) -> float:
        if len(self.path) <= 1:
            return 1.0 if self.state == AgentState.ARRIVED else 0.0
        return self.path_index / (len(self.path) - 1)

    # ── Serialization ─────────────────────────────────────────────────────

    def to_dict(self) -> dict:
        return {
            "id":       self.id,
            "state":    self.state.value,
            "grid_x":   self.grid_x,
            "grid_y":   self.grid_y,
            "goal_x":   self.goal["x"],
            "goal_y":   self.goal["y"],
            "priority": self.priority,
            "progress": round(self.progress, 3),
            "steps":    self.steps_taken,
            "waits":    self.wait_steps,
            "replans":  self.replan_count,
        }

    def reset(self, new_start: dict, new_goal: dict, priority: int) -> None:
        self.__init__(self.id, new_start, new_goal, priority)

File: multi-agent-pathfinding/js/agent.js
/**
 * agent.js — Agent Class
 * Each agent has a unique ID, start/goal positions, a planned space-time path,
 * and a state machine driving its behaviour.
 *
 * State Machine:
 *   IDLE → PLANNING → MOVING ↔ WAITING → ARRIVED
 *                  ↘ STUCK (if no path found)
 */

'use strict';

const AgentState = Object.freeze({
  IDLE:     'IDLE',
  PLANNING: 'PLANNING',
  MOVING:   'MOVING',
  WAITING:  'WAITING',
  ARRIVED:  'ARRIVED',
  STUCK:    'STUCK',
});

class Agent {
  /**
   * @param {number} id      - unique integer id
   * @param {{x,y}} start    - starting grid cell
   * @param {{x,y}} goal     - target grid cell
   * @param {number} priority - lower number = higher priority during PBS
   */
  constructor(id, start, goal, priority) {
    this.id       = id;
    this.start    = { ...start };
    this.goal     = { ...goal  };
    this.priority = priority;

    // Current discrete grid position (updated each timestep)
    this.gridX = start.x;
    this.gridY = start.y;

    // Render interpolation (pixel-space lerp between grid steps)
    this.renderX = start.x;
    this.renderY = start.y;

    // Planned space-time path: [{x, y, t}, ...]
    // Index 0 is start at t=startTime, last is goal.
    this.path = [];

    // Index into path array for the current simulation timestep
    this.pathIndex = 0;

    // Current simulation timestep at which agent was placed
    this.startTime = 0;

    this.state = AgentState.IDLE;
    this.replanCount = 0;

    // Colour (HSL hue, 0–360) — assigned externally for visual differentiation
    this.hue = (id * 137.5) % 360; // golden-angle spacing

    // Statistics
    this.stepsTaken = 0;
    this.waitSteps  = 0;
    this.conflictsInvolved = 0;

    // Trail of recent grid positions for rendering
    this._trail = [];
    this._trailMaxLen = 12;
  }

  // -------------------------------------------------------------------------
  // Path management
  // -------------------------------------------------------------------------

  /** Assign a newly-planned space-time path. */
  assignPath(path) {
    this.path = path ?? [];
    this.pathIndex = 0;
    this.state = this.path.length > 0 ? AgentState.MOVING : AgentState.STUCK;
  }

  /** True if agent still has steps to take. */
  hasRemainingPath() {
    return this.pathIndex < this.path.length - 1;
  }

  /**
   * Advance agent by one simulation timestep.
   * @param {number} currentT - the current global timestep
   */
  step(currentT) {
    if (this.state === AgentState.ARRIVED || this.state === AgentState.STUCK) return;
    if (this.path.length === 0) { this.state = AgentState.STUCK; return; }

    // Find the path entry for this timestep
    const idx = this.pathIndex;
    if (idx >= this.path.length) {
      this.state = AgentState.ARRIVED;
      return;
    }

    const entry = this.path[idx];
    if (entry.t > currentT) {
      // Agent hasn't started yet (shouldn't happen in normal flow)
      this.state = AgentState.WAITING;
      return;
    }

    // Move to the cell in the path at or before currentT
    const newX = entry.x;
    const newY = entry.y;

    const moved = (newX !== this.gridX || newY !== this.gridY);
    if (moved) {
      this._trail.push({ x: this.gridX, y: this.gridY });
      if (this._trail.length > this._trailMaxLen) this._trail.shift();
      this.stepsTaken++;
    } else {
      this.waitSteps++;
    }

    this.gridX = newX;
    this.gridY = newY;

    // Advance path index for next timestep
    if (idx + 1 < this.path.length) {
      this.pathIndex++;
    }

    // Check arrival
    if (this.gridX === this.goal.x && this.gridY === this.goal.y) {
      this.state = AgentState.ARRIVED;
    } else {
      this.state = moved ? AgentState.MOVING : AgentState.WAITING;
    }
  }

  /**
   * Update render interpolation position.
   * @param {number} alpha - 0=prev position, 1=current grid position
   */
  updateRenderPosition(alpha) {
    this.renderX += (this.gridX - this.renderX) * alpha;
    this.renderY += (this.gridY - this.renderY) * alpha;
  }

  /** Snap render position immediately (used on reset). */
  snapRenderPosition() {
    this.renderX = this.gridX;
    this.renderY = this.gridY;
  }

  /** Reset agent to initial conditions (for scenario reset). */
  reset(newStart, newGoal, priority) {
    this.start    = { ...newStart };
    this.goal     = { ...newGoal  };
    this.priority = priority;
    this.gridX    = newStart.x;
    this.gridY    = newStart.y;
    this.renderX  = newStart.x;
    this.renderY  = newStart.y;
    this.path     = [];
    this.pathIndex = 0;
    this.startTime = 0;
    this.state    = AgentState.IDLE;
    this.replanCount = 0;
    this.stepsTaken  = 0;
    this.waitSteps   = 0;
    this.conflictsInvolved = 0;
    this._trail   = [];
  }

  get trail() { return this._trail; }

  /** Progress 0→1 along path. */
  get progress() {
    if (this.path.length <= 1) return this.state === AgentState.ARRIVED ? 1 : 0;
    return this.pathIndex / (this.path.length - 1);
  }
}
